pkg/agent/loadbalancer/loadbalancer_test.go

```diff
@@ -82,221 +91,366 @@ func ping(conn net.Conn) (string, error) {
-		t.Skip("skipping slow test in short mode.")
-	}
-	tmpDir := t.TempDir()
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	serverAddr := "192.0.2.1:6443"
-	lb, err := New(ctx, tmpDir, SupervisorServiceName, "http://"+serverAddr, RandomPort, false)
-	if err != nil {
-		t.Fatalf("New() failed: %v", err)
-	}
-
-	// Set failing health check to reduce retries
-	lb.SetHealthCheck(serverAddr, func() bool { return false })
-
-	conn, err := net.Dial("tcp", lb.localAddress)
-	if err != nil {
-		t.Fatalf("net.Dial failed: %v", err)
-	}
-
-	done := make(chan error)
-	go func() {
-		_, err = ping(conn)
-		done <- err
-	}()
-	timeout := time.After(11 * time.Second)
-
-	select {
-	case err := <-done:
-		if err == nil {
-			t.Fatal("Unexpected successful ping from unreachable address")
-		}
-	case <-timeout:
-		t.Fatal("Test timed out")
-	}
-}
+var _ = Describe("LoadBalancer", func() {
+	// creates a LB using a default server (ie fixed registration endpoint)
+	// and then adds a new server (a node). The node server is then closed, and it is confirmed
+	// that new connections use the default server.
+	When("loadbalancer is running", Ordered, func() {
+		ctx, cancel := context.WithCancel(context.Background())
+		var defaultServer, node1Server, node2Server *testServer
+		var conn1, conn2, conn3, conn4 net.Conn
+		var lb *LoadBalancer
+		var err error
+
+		BeforeAll(func() {
+			tmpDir := GinkgoT().TempDir()
+
+			defaultServer, err = createServer(ctx, "default")
+			Expect(err).NotTo(HaveOccurred(), "createServer(default) failed")
+
+			node1Server, err = createServer(ctx, "node1")
+			Expect(err).NotTo(HaveOccurred(), "createServer(node1) failed")
+
+			node2Server, err = createServer(ctx, "node2")
+			Expect(err).NotTo(HaveOccurred(), "createServer(node2) failed")
+
+			// start the loadbalancer with the default server as the only server
+			lb, err = New(ctx, tmpDir, SupervisorServiceName, "http://"+defaultServer.address(), RandomPort, false)
+			Expect(err).NotTo(HaveOccurred(), "New() failed")
+		})
+
+		AfterAll(func() {
+			cancel()
+		})
+
+		It("adds node1 as a server", func() {
+			// add the node as a new server address.
+			lb.Update([]string{node1Server.address()})
+			lb.SetHealthCheck(node1Server.address(), func() HealthCheckResult { return HealthCheckResultOK })
+
+			By(fmt.Sprintf("Added node1 server: %v", lb.servers.getServers()))
+
+			// wait for state to change
+			Eventually(func() state {
+				if s := lb.servers.getServer(node1Server.address()); s != nil {
+					return s.state
+				}
+				return stateInvalid
+			}, 5, 1).Should(Equal(statePreferred))
+		})
+
+		It("connects to node1", func() {
+			// make sure connections go to the node
+			conn1, err = net.Dial("tcp", lb.localAddress)
+			Expect(err).NotTo(HaveOccurred(), "net.Dial failed")
+
+			result, err := ping(conn1)
+			Expect(err).NotTo(HaveOccurred(), "ping(conn1) failed")
+			Expect(result).To(Equal("node1:ping"), "Unexpected ping(conn1) result")
```

Comment:
Nit/Personal Preference. You can compress this into a single line statement. This is nice when you don't need/keep the return values for later.
```suggestion
			Expect(ping(conn1)).To(Equal("node1:ping"), "Unexpected ping(conn1) result")
```

Gomega handles multiple return values, the first is passed to the matcher and the rest (ie err in this case) are check to be nil